### DISCUSSION TITLE
Model examples with Early Stopping breaking one-class-classification assumption?

### DISCUSSION THREAD

**Discussion by DeclanMcIntosh** [open-edge-platform/anomalib#3317] [General]
The YAML default configs for the models often use early stopping as below, based on image_AUROC or pixel_AUROC. Is this not using the labelled data and breaking the OCC assumption of the original method? Would this not make these methods rely on labelled test or validation data of both nominal and anomalous structures to replicate the results?  

Example from examples/configs/model/reverse_distillation.yaml

```
trainer:
  callbacks:
    - class_path: lightning.pytorch.callbacks.EarlyStopping
      init_args:
        patience: 3
        monitor: pixel_AUROC
        mode: max
  check_val_every_n_epoch: 200
  max_epochs: 200
```

**Comment by jgespino** [comment]
@samet-akcay Could you please help assign this discussion to the team?

**Comment by ashwinvaidya17** [comment]
@DeclanMcIntosh unfortunately that is still an open problem in this field. It affects the threshold selection as well. We use max F1 score to select the pixel-level and image-level thresholds. F1 score needs "anomaly" images from the validation set. The _single-label_ method to go for now would be to use something like a proxy signal for anomalous labels. Anomalib supports synthetic anomaly generation. So you can set the val split to synthetic (https://anomalib.readthedocs.io/en/latest/markdown/guides/reference/data/utils/image_video.html#anomalib.data.utils.split.ValSplitMode.SYNTHETIC).

Other option would be to use just the loss for early stopping. The threshold would then be the maximum score on the normal images
https://github.com/open-edge-platform/anomalib/blob/0a6e97b7497d661cc684bb3616086250f1263fb6/src/anomalib/metrics/threshold/f1_adaptive_threshold.py#L124

**Reply by DeclanMcIntosh** [reply]
@ashwinvaidya17 thanks for the reply, I am currently working on a method to make OCC methods fully unsupervised and was trying to figure out the best practice for this! I found that early stopping led to poor performance for some methods, and that simply removing early stopping and training for the default maximum epochs or steps worked better in general. In case anyone is looking to do something similar in the future. 

I had not seen the synthetic splits, so I will definitely look into that! Thanks for your help! 

Cheers!

**Reply by abc-125** [reply]
> I found that early stopping led to poor performance for some methods, and that simply removing early stopping and training for the default maximum epochs or steps worked better in general.

@DeclanMcIntosh it is interesting, can you please tell me which methods showed worse performance? We had a paper last year that concluded early stopping (even if it is not on a test set) using image_AUROC [improves results](https://arxiv.org/abs/2503.23451), but we had a rather small selection of methods.

**Reply by DeclanMcIntosh** [reply]
@abc-125  I was using early stopping based on the training loss since we were doing fully unsupervised training. We tested this with Dinomaly, Dfm, Fastflow, and Padim. But I did not do a thorough analysis of this; I just moved on quickly to training for the entire period when that showed better results. Unfortunately, I don't even have the before-and-after results for each of these recorded. I would not be surprised if using other metrics for early stopping performs better.
